Source/Core/sampleTerrain.js

```diff
@@ -48,6 +48,78 @@ function sampleTerrain(terrainProvider, level, positions) {
   });
 }
 
+const ACTION__CONTINUE = "continue";
+const ACTION__DELAY = "delay";
+
+function consumeQueueItem(tileRequests, inflightRequests) {
+  const tileRequest = tileRequests[0];
+  const requestPromise = tileRequest.terrainProvider.requestTileGeometry(
+    tileRequest.x,
+    tileRequest.y,
+    tileRequest.level
+  );
+
+  if (!requestPromise) {
+    // Uh, we probably got throttled client side?
+    //  let's wait 100ms, then try again
+    // This will turn into an infinite loop if we don't eventually
+    //  get allowed through...
+    return Promise.resolve(ACTION__DELAY);
+  }
+
+  const promise = requestPromise
+    .then(createInterpolateFunction(tileRequest))
+    .catch(createMarkFailedFunction(tileRequest));
+
+  // remove the request we've just done from the queue
+  //  and add its promise result to the inflight request list
+  tileRequests.shift();
+  inflightRequests.push(promise);
+
+  // our next action is to do the next item in the queue
+  return Promise.resolve(ACTION__CONTINUE);
+}
+
+function delay(ms) {
+  return new Promise(function (res) {
+    window.setTimeout(res, ms);
+  });
+}
+
+function log(msg) {
+  console.log(msg);
+}
+
+function drainTileRequestQueue(tileRequests, results) {
+  log(
+    `draining request queue, there are ${tileRequests.length} to go and ${results.length} in flight or complete`
+  );
+  // nothing to do, return
+  if (!tileRequests.length) {
+    return Promise.resolve(null);
```

Comment:
We use `undefined` instead of `null`, so this can be simplified to `Promise.resolve()`.